### DISCUSSION TITLE
Lazy+async: can't we just return awaitables where we currently throw exceptions?

### DISCUSSION THREAD

**Discussion by t184256** [sqlalchemy/sqlalchemy#9731] [Ideas]
[Documentation](https://docs.sqlalchemy.org/en/20/orm/extensions/asyncio.html#asyncio-orm-avoid-lazyloads) seems insistent, that with async code I should do one of the following:

* eaqerly load everything (recommended for some reason)
* write code to pass to [`run_sync`](https://docs.sqlalchemy.org/en/20/orm/extensions/asyncio.html#sqlalchemy.ext.asyncio.AsyncSession.run_sync)
* `await async_session.refresh(a1, ["bs"])`

Accessing a lazy relationship raises an exception.

Question: why not just return an awaitable for lazy-loaded attributes instead of those exceptions? This way one would be able to `await a1.bs` (in terms of [async_orm example](https://docs.sqlalchemy.org/en/20/_modules/examples/asyncio/async_orm.html)) to perform explicit loading, on-demand, with just the standard keyword that IO is hiding behing?

Here's a quickly hacked together mix-in to illustrate what I have in mind:

```python
class AsyncLoadableRelationshipsMixin:
    def __getattribute__(self, name):
        cls = object.__getattribute__(self, '__class__')
        relationships = sqlalchemy.inspection.inspect(cls).relationships
        annotations = object.__getattribute__(self, '__annotations__')
        if name in relationships and name in annotations:
            if isinstance(relationships[name].strategy,
                          sqlalchemy.orm.strategies.LazyLoader):
                # lazy-loaded attribute, construct an awaitable returning it
                if name in sqlalchemy.inspection.inspect(self).unloaded:
                    # loading is needed
                    session = sqlalchemy.orm.Session.object_session(self)
                    async def loader():
                        await session.refresh(self, attribute_names=[name])
                        return object.__getattribute__(self, name)
                else:
                    # loading is not neeeded, but return an awaitable anyway
                    # to have the interface uniform
                    async def loader():
                        return object.__getattribute__(self, name)
                return loader()
        return object.__getattribute__(self, name)
```

Replace Base in the example with:
``` python
class Base(sqlalchemy.orm.DeclarativeBase, AsyncLoadableRelationshipsMixin):
    pass
```

and you'll be able to sprinkle `await`s around accessors and thus load attributes with just `await a1.bs` and all potential IO being explicit. Maybe even return awaitables which are also async iterators, but that's details.

Why not return awaitables for lazy relationships? In `lazy` strategy or, if compatibility is a concern, in some `lazy-awaitable` strategy.

---

My attempt at summarizing the discussion below:
1. any attribute on the class can emit IO => making every attribute access return an awaitable changes the way every attribute is accessed => too invasive of a change
2. returning dual-purpose objects that act as both awaitables and values would be too big of a change as well
3. it might be possible to implement an async decorator that'd let one `async with run_sync():` and use familiar sync syntax without the need to define and pass functions
4. `async with a1.something_something.bs` is another option, and an easy to implement one

**Comment by zzzeek** [comment]
stylistically it strikes me as poor.  I'd prefer a library function that makes it explicit this action is running IO.

    await invoke_attribute(a1.bs)

**Reply by zzzeek** [reply]
lazy loading would still need to return a custom object however any solution we do here would need to be generalized to **all** expired attributes, deferreds, etc.  otherwise it's just a hack

**Reply by zzzeek** [reply]
which gets into that it's very difficult to make this work actually calling a1.bs ahead of the await or function.   We can add right now, without changing anything at all, this function instead:

    await invoke_attribute(lambda: a1.bs)

we have that now.  it's just the lambda that's unpleasant.

**Reply by zzzeek** [reply]
or perhaps we can do it like this:

```py
async with run_sync():
   a1.bs
```

how about that?

**Reply by zzzeek** [reply]
overall, the reason none of this is "there" is due to the "controversy" of SQLAlchemy's asyncio approach, that dares mix asyncio with sync code.   But it's been almost three years now and the world didnt end and Guido hasn't arrived with a flamethrower, so maybe it's time.

**Reply by zzzeek** [reply]
having a1.bs return an awaitable directly, I dont see how that really would work.  once a1.bs is lazy-loaded, it's a plain Python `list`.  You can't await that.    Trying to ask users to always know if `a1.bs` is loaded or not would not be feasible

**Reply by t184256** [reply]
Not sure I share this feeling, I'd argue it's the act of `await`ing in itself that's the explicit and canonical way of yielding control to something that'd do IO. If `await`'s argument is made to look as a function call, that actually feels misleading to me.

**Reply by zzzeek** [reply]
I dont see how a1.bs could be awaitable in only some cases.    if you think it through you will see this is not feasible.

**Reply by t184256** [reply]
> async with run_sync():
>    a1.bs

That's shorter, but conceals where does the IO happen and looks as some no-op at a first glance. Is this even possible to implement it as an async context manager?

> having a1.bs return an awaitable directly, I dont see how that really would work. once a1.bs is lazy-loaded, it's a plain Python list. You can't await that. Trying to ask users to always know if a1.bs is loaded or not would not be feasible.

True, that's why my example returns awaitables regardless of whether the attribute is loaded or not.

Alternatively, some custom collection can be returned, that can't be meaningfully accessed until it's awaited.

**Reply by zzzeek** [reply]
please keep in mind as well, **all attributes on the mapped object can emit IO**.   So this would be an API that looks like this:

```py

obj = MyObj(a=1, b=2)
a = await obj.a
b = await obj.b
```

i.e. it would be fully unworkable.  all methods on the object would need to be `async` and it would be impossible to refer to instance variables in methods like ``__init__()``, ``__repr__()``, etc.  pep-484 support would be gone.   the internals would need massive changes to accommodate this and it's probably not even possible.  

So a function or context manager that wraps the attribute access is fully necessary.

**Reply by t184256** [reply]
>  please keep in mind as well, all attributes on the class can emit IO

Eh? It's only the lazy loaded ones being denied access to, the others just work, so I assume it is fine now and there's no need to change them. I proposed to only return awaitables for lazy-loaded ones, which one already can't straightforwardly access without explicit preloading on the side. Not return awaitables for every single attribute.

**Reply by zzzeek** [reply]
> > async with run_sync():
> > a1.bs
> 
> That's shorter, but conceals where does the IO happen and looks as some no-op at a first glance. Is this even possible to implement it as an async context manager?

we have a system that runs awaitables without the "await" keyword.  So the context manager would plug that in, likely using a context.local variable, and our await_() function would have to call into this context.     it's a lot cleaner to pass a callable to run_sync(), though.

**Reply by t184256** [reply]
> > > `async with run_sync(): a1.bs`
> > 
> > 
> > That's shorter, but conceals where does the IO happen and looks as some no-op at a first glance. Is this even possible to implement it as an async context manager?
> 
> we have a system that runs awaitables without the "await" keyword. So the context manager would plug that in, likely using a context.local variable, and our await_() function would have to call into this context. it's a lot cleaner to pass a callable to run_sync(), though.

Well, if it's doable, feel free to make an RFE out of this idea.

But I still want to squeeze extra usefulness from the current arrangement specifically in cases where one currently just cannot freely `a1.bs` at all.

**Reply by zzzeek** [reply]
> please keep in mind as well, all attributes on the class can emit IO

> Eh? It's only the lazy loaded ones being denied access to, the others just work, 

that's not correct.   try using a deferred attribute, using joined table inheritance and accessing an attribute local to a subclass, or calling session.expire() on your object.

When I said "any attribute on the class can emit IO" I wasn't just making that up.  I've been here since 2006....

**Reply by t184256** [reply]
Oh, I see now.

I'd still prefer an awaitable, `await`ing on which "just does what I mean" over exceptions that make me dive in the docs, frown through recommendations to do other things instead, and only then figuring out the exact `await something_else` that'd give me what I originally wanted to have. Context manager that magically sets things right is a step in that direction, though I'd rather prefer a bland `await` I find much harder to misinterpret.

**Reply by zzzeek** [reply]
for some context, the entire notion that an ORM like SQLAlchemy, strongly oriented around the lazy loading pattern, could ever work with Python asyncio was basically a non-starter.  This whole thing is already well within the realm of **completely impossible**. 

The asyncio extension was added to show that basically all of SQLAlchemy Core, and *a large portion* of the ORM, are actually *possible* to use with asyncio.  the extension was developed and produced with the caveat though that the lazy loading pattern is and always will be fundamentally incompatible with asyncio.  That's basically the arrangement.

It's just been three years now, and basically every Python library being made now defaults to "asyncio", and people are forgetting, this whole thing, is supposed to be *impossible*.     We made it possible, but there's a caveat, that you can't have attributes load on demand.    Not without some other "thing", some magic funciton, context, some magic attribute, etc. that is explicit.

Even this would be more doable:

    x = await myobject.async_attrs.bs

that would be doable.   Also fairly easily, that could just be a simple mixin class.   wouldn't have typing info though...

**Comment by zzzeek** [comment]
Here's async_attrs.  This is extremely doable, can be added to sqlalchemy.ext.asyncio in 5 minutes and we're done with it.  The whole section at https://docs.sqlalchemy.org/en/20/orm/extensions/asyncio.html#preventing-implicit-io-when-using-asyncsession could be replaced at the top with a single paragraph to use this extension; then have all the other bullets afterwards as other ways to avoid the same issue:

```py
from __future__ import annotations

import asyncio
import datetime
from typing import Any
from typing import Awaitable
from typing import List, TypeVar
from typing import Optional

from sqlalchemy import ForeignKey
from sqlalchemy import func
from sqlalchemy.ext.asyncio import async_sessionmaker
from sqlalchemy.ext.asyncio import create_async_engine
from sqlalchemy.future import select
from sqlalchemy.orm import DeclarativeBase
from sqlalchemy.orm import Mapped
from sqlalchemy.orm import mapped_column
from sqlalchemy.orm import relationship
from sqlalchemy.util import greenlet_spawn


_T = TypeVar("_T", bound=Any)


class AwaitAttrs:
    class _AwaitAttrGetitem:
        __slots__ = "_instance"

        def __init__(self, _instance: Any):
            self._instance = _instance

        def __getattr__(self, name: str) -> Awaitable[Any]:
            return greenlet_spawn(getattr, self._instance, name)

    @property
    def await_attr(self) -> AwaitAttrs._AwaitAttrGetitem:
        """provide awaitable attribute access"""
        return AwaitAttrs._AwaitAttrGetitem(self)

    async def await_load(self, attr: Mapped[_T]) -> _T:
        """typed version of getattr"""
        return await greenlet_spawn(
            getattr, self, attr.key  # type: ignore
        )


class Base(AwaitAttrs, DeclarativeBase):
    pass


class A(Base):
    __tablename__ = "a"

    id: Mapped[int] = mapped_column(primary_key=True)
    data: Mapped[Optional[str]]
    create_date: Mapped[datetime.datetime] = mapped_column(
        server_default=func.now()
    )
    bs: Mapped[List[B]] = relationship()


class B(Base):
    __tablename__ = "b"
    id: Mapped[int] = mapped_column(primary_key=True)
    a_id: Mapped[int] = mapped_column(ForeignKey("a.id"))
    data: Mapped[Optional[str]]


async def async_main():

    engine = create_async_engine(
        "postgresql+asyncpg://scott:tiger@localhost/test",
        echo=True,
    )

    async with engine.begin() as conn:
        await conn.run_sync(Base.metadata.drop_all)
    async with engine.begin() as conn:
        await conn.run_sync(Base.metadata.create_all)

    async_session = async_sessionmaker(engine)

    async with async_session() as session:
        async with session.begin():
            session.add_all(
                [
                    A(bs=[B(), B()], data="a1"),
                ]
            )

        stmt = select(A).limit(1)

        a1 = (await session.scalars(stmt)).one()

        for b1 in await a1.await_attr.bs:
            print(b1)

        # a typed version
        for b1 in await a1.await_load(A.bs):
            print(b1)

        a1.data = "new data"

        await session.commit()

        unexpired_data = await a1.await_attr.data
        print(f"unexpired data: {unexpired_data}")


asyncio.run(async_main())

```

**Reply by t184256** [reply]
> AwaitAttrs/.await_

AsyncAttrs/.async_, maybe? So that we don't await for await-something, but await- for async-something.

> implementation and interface

Sorry, I don't feel knowledgeable enough about the whole thing to judge whether it's the best thing since sliced bread or there's something more to improve on. So I'll wear the only hat available to me, a newbie user hat, and just state that `await a1.some_magic.data` does feel cleaner to me than both `await async_session.refresh(a1, names=["bs"])` and/or `async with run_sync(): a1.bs`, especially since it's an expression I can directly use somewhere.

**Reply by CaselIT** [reply]
>  ```py
>  for b1 in await a1.await_load(A.bs):
>            print(b1)
>  ```

Not sure I like this, since users will try to do `a1.await_load(a1.bs)` that will raise.
I very much would like to add only the non-typed version and point users towards the typing repo regarding adding support for syntax. The issue to follow seems this https://github.com/python/typing/issues/685 and the more recent https://github.com/python/typing/discussions/1387

**Reply by zzzeek** [reply]
i didnt really follow those issues, but absolutely, dont add the typed version right now.    I mean dont add any of this :)  unless we think it's not just adding fuel to the fire

**Reply by zzzeek** [reply]
> > AwaitAttrs/.await_
> 
> AsyncAttrs/.async_, maybe? So that we don't await for await-something, but await- for async-something.

well it's an "awaitable", i will start with the most accurate name `awaitable_attrs` and we can change / add synonyms from there

**Reply by t184256** [reply]
Can't argue with `awaitable_attrs`.

**Reply by zzzeek** [reply]
well your fingers typing all those letters will...

**Reply by zzzeek** [reply]
https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4600
